crates/network-tunnel/src/main.rs

```diff
@@ -22,21 +22,129 @@ async fn main() -> io::Result<()> {
     Ok(())
 }
 
+async fn run_and_cleanup(tunnel: &mut Box<dyn networktunnel::NetworkTunnel>) -> Result<(), Error> {
+    let tunnel_block = {
+        let prep = tunnel.prepare().await;
+
+        // Write "READY" to stdio to unblock Go logic.
+        // The current workflow assumes that
+        //   1. After tunnel.prepare() is called, the network tunnel is able to accept requests from clients without sending errors back to clients.
+        //   2. The network tunnel is able to process client requests immediately after `tunnel.start_serve` is called.
+        // If either of the assumptions is invalid for any new tunnel type, the READY-logic need to be moved to a separate task, which
+        //    sends out the "READY" signal after making sure the network tunnel is started and working properly.
+        println!("READY");
+
+        prep.and(tunnel.start_serve().await)
```

Comment:
So as written here, I think this will run `tunnel.start_serve()` even if the result of `tunnel.prepare()` is an error. Is this the intended behavior, or should this maybe be `prep.and_then()` so it skips the `start_serve()` when prepare resulted in an error?

Reply:
@willdonnelly great catch, you are right it should be `and_then`